### ISSUE TITLE
Simplification of unions involving type variables

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#1666]
Consider the following example:
```py
from typing import TypeVar, Any

type IntOr[T: int] = int | T

def _(x: IntOr[Any]):
    reveal_type(x)
```
https://play.ty.dev/b378c2a3-d007-428d-8d59-a3a69f5614c0

ty reveals `int` here, but every other type checker reveals `int | Any`, which seems more reasonable? ty eagerly simplifies the `int | T` union to `int`, which is correct for every static `T <: int`... but leads to surprising results when `T` is explicitly specialized to a dynamic type.

This was prompted by a much more complex example in numpy's codebase:

```py
class _SupportsDType(Protocol[_DTypeT_co]):
    @property
    def dtype(self) -> _DTypeT_co: ...

_DTypeLike = type[_ScalarT] | _SupportsDType[dtype[_ScalarT]]
```
`_ScalarT` has an upper bound of `np.generic` (= `np.generic[Any]`). And `np.generic[…]` does have a `dtype` property member. This currently leads us to treat `type[_Scalar]` as a subtype of `_SupportsDType[…]`, and so that union gets simplified to `_SupportsDType[dtype[_ScalarT]]`. Later, `_DTypeLike` is explicitly specialized with `_DTypeLike[Any]`, and it is expected that `<class 'object'>` should be assignable to `_DTypeLike[Any]` (which would be the case if `type[Any]` were still part of the union.

**Comment by carljm** [comment]
Yes, it seems to me we probably need to account for the possibility of specialization to a dynamic type, and avoid doing this kind of simplification. Which would be achieved by considering `T: int` to be assignable-to, but not subtype-of or redundant-with, `int` (or `object`).

**Comment by AlexWaygood** [comment]
> Which would be achieved by considering `T: int` to be assignable-to, but not subtype-of or redundant-with, `int` (or `object`).

As discussed in https://github.com/astral-sh/ty/issues/1846#issuecomment-3639019092, I agree that we probably shouldn't automatically treat `TypeVar`s as subtypes of (or redundant with) their upper bound. But I do think that we should continue to see all `TypeVar`s as subtypes of (and redundant with) `object`, because _all_ types (even `Any`!) are subtypes of `object`. So accounting for the possibility that a `TypeVar` might be solved to `Any` is immaterial there.

**Comment by charliermarsh** [comment]
There is a narrow version of this here: https://github.com/astral-sh/ruff/pull/26118. But it's not the full change that @AlexWaygood suggested above.

**Comment by carljm** [comment]
This seems questionable, on a second look. If a typevar `T` has upper bound `S`, and is explicitly specialized to `Any`, shouldn't we ideally consider that to be `S & Any` due to the upper bound? This is what is effectively suggested by #1109 and related issues. It seems correct to me for `IntOr[Any]` to be `int` rather than `int | Any`; the latter ignores the upper bound. If that behavior isn't desired, then `T` should not have the upper bound `int`.

I guess this issue should be summarized as "can dynamic specializations be assumed to ignore upper bounds"?

The counter argument is that in general dynamic types are unsound, and we don't (in general) collect constraints on a given dynamic type to ensure they are all consistent with each other. So maybe dynamic specializations should be no exception.

**Comment by carljm** [comment]
I don't think we need to change anything here.
